multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactoryTest.java

```diff
@@ -0,0 +1,100 @@
+import static org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProjectTest.scheduleAndFindBranchProject;
+import org.jenkinsci.plugins.workflow.steps.scm.GitSampleRepoRule;
+import org.junit.Test;
+import static org.junit.Assert.*;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+public class WorkflowBranchProjectFactoryTest {
+
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+    @Rule public GitSampleRepoRule sampleRepo = new GitSampleRepoRule();
+
+    @Issue("JENKINS-30744")
+    @Test public void slashyBranches() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                sampleRepo.init();
+                sampleRepo.git("checkout", "-b", "dev/main");
+                ScriptApproval.get().approveSignature("method java.lang.String replaceFirst java.lang.String java.lang.String"); // TODO add to generic-whitelist
+                sampleRepo.write("Jenkinsfile",
+                    "echo \"branch=${env.BRANCH_NAME}\"\n" +
+                    "node {\n" +
+                    "  checkout scm\n" +
+                    "  echo \"workspace=${pwd().replaceFirst('.+dev', 'dev')}\"\n" +
+                    "}");
+                sampleRepo.git("add", "Jenkinsfile");
+                sampleRepo.git("commit", "--all", "--message=flow");
+                WorkflowMultiBranchProject mp = story.j.jenkins.createProject(WorkflowMultiBranchProject.class, "p");
+                mp.getSourcesList().add(new BranchSource(new GitSCMSource(null, sampleRepo.toString(), "", "*", "", false), new DefaultBranchPropertyStrategy(new BranchProperty[0])));
+                WorkflowJob p = scheduleAndFindBranchProject(mp, "dev%2Fmain");
+                assertEquals(1, mp.getItems().size());
+                story.j.waitUntilNoActivity();
+                WorkflowRun b1 = p.getLastBuild();
+                assertEquals(1, b1.getNumber());
+                story.j.assertLogContains("branch=dev/main", b1);
+                story.j.assertLogContains("workspace=dev%2Fmain", b1);
+                verifyProject(p);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p/dev%2Fmain", WorkflowJob.class);
+                assertNotNull(p);
+                WorkflowRun b2 = p.scheduleBuild2(0).get();
+                assertEquals(2, b2.getNumber());
+                story.j.assertLogContains("branch=dev/main", b2);
+                story.j.assertLogContains("workspace=dev%2Fmain", b2);
+                verifyProject(p);
+            }
+        });
+    }
```

Comment:
Lack of newline intentional, since the method is just a helper only used in the method above.

If Java allowed local functions I would use them. If we were in `-source 8` for `src/test/java/` I would perhaps use lambdas, though it is awkward since Java still lacks anonymous SAMs, and the standard `Consumer` does not take checked exceptions, and Guava has no apparent replacement, so you are stuck with a local SAM definition plus lambdas:

```java
@Test public void whatever() {
    interface Verifier {void verifyProject(WorkflowJob p) throws Exception;}
    final Verifier verifier = (p) -> {
        // …
    };
    // …
    verifier.verifyProject(p);
}
```

which seems more trouble than it is worth.